README.md
[![Build Status](https://travis-ci.org/TRIQS/ctint_tutorial.svg?branch=unstable)](https://travis-ci.org/TRIQS/ctint_tutorial)

# ctint_tutorial - A skeleton for a TRIQS application

Initial Setup
-------------

To adapt this skeleton for a new TRIQS application, the following steps are necessary:

* Create a repository, e.g. https://github.com/username/appname

* Run the following commands in order after replacing **appname** accordingly

```bash
git clone https://github.com/triqs/ctint_tutorial --branch unstable appname
cd appname
./share/squash_history.sh
./share/replace_and_rename.sh appname
git add -A && git commit -m "Adjust ctint_tutorial skeleton for appname"
```

You can now add your github repository and push to it

```bash
git remote add origin https://github.com/username/appname
git remote update
git push origin unstable
```

If you prefer to use the [SSH interface](https://help.github.com/en/articles/connecting-to-github-with-ssh)
to the remote repository, replace the http link with e.g. `git@github.com:username/appname`.

### Merging ctint_tutorial skeleton updates ###

You can merge future changes to the ctint_tutorial skeleton into your project with the following commands

```bash
git remote update
git merge ctint_tutorial_remote/unstable -m "Merge latest ctint_tutorial skeleton changes"
```

If you should encounter any conflicts resolve them and `git commit`.
Finally we repeat the replace and rename command from the initial setup.

```bash
./share/replace_and_rename.sh appname
git commit --amend
```

Getting Started
---------------

After setting up your application as described above you should customize the following files and directories
according to your needs (replace ctint_tutorial in the following by the name of your application)

* Adjust or remove the `README.md` and `doc/ChangeLog.md` file
* In the `c++/ctint_tutorial` subdirectory adjust the example files `ctint_tutorial.hpp` and `ctint_tutorial.cpp` or add your own source files.
* In the `test/c++` subdirectory adjust the example test `basic.cpp` or add your own tests.
* In the `python/ctint_tutorial` subdirectory add your Python source files.
  Be sure to remove the `ctint_tutorial_module_desc.py` file unless you want to generate a Python module from your C++ source code.
* In the `test/python` subdirectory adjust the example test `Basic.py` or add your own tests.
* Adjust any documentation examples given as `*.rst` files in the doc directory.
* Adjust the sphinx configuration in `doc/conf.py.in` as necessary.
* The build and install process is identical to the one outline [here](https://triqs.github.io/ctint_tutorial/unstable/install.html).

### Optional ###
----------------

* If you want to wrap C++ classes and/or functions provided in the `c++/ctint_tutorial/ctint_tutorial.hpp` rerun the `c++2py` tool with
```bash
c++2py -r ctint_tutorial_module_desc.py
```
* Add your email address to the bottom section of `Jenkinsfile` for Jenkins CI notification emails
```
End of build log:
\${BUILD_LOG,maxLines=60}
    """,
    to: 'user@domain.org',
```


doc/changelog.rst
.. _changelog:

Changelog
=========

This document describes the main changes in CTINT_TUTORIAL.

.. include:: ChangeLog.md


test/c++/basic.cpp
#include <triqs/test_tools/gfs.hpp>
#include <ctint_tutorial/ctint_tutorial.hpp>

using namespace ctint_tutorial;

TEST(Toto, Add) { // NOLINT

  toto a(0);
  toto b(2);

  auto c = a + b;
  EXPECT_EQ(c, b); // NOLINT
}

TEST(Toto, H5) { // NOLINT

  toto a(0);
  { // Local scope for file
    triqs::h5::file f("f.h5", H5F_ACC_TRUNC);
    h5_write(f, "a", a);
  }

  toto a2;
  {
    triqs::h5::file f("f.h5", H5F_ACC_RDWR);
    h5_read(f, "a", a2);
  }

  EXPECT_EQ(a, a2); // NOLINT
}


test/python/Basic.py
#!/usr/bin/env python

import unittest

from ctint_tutorial import Toto, chain
from pytriqs.archive import *
from pytriqs.utility import mpi

class test_toto(unittest.TestCase):

    def test_add(self):

        a=Toto(0)
        b=Toto(2)
        
        c=a+b
        self.assertEqual(c, b)


    def test_h5(self):
        
        a=Toto(0)
        with HDFArchive("f.h5",'w') as A:
            A["a"] = a
        with HDFArchive("f.h5",'r') as A:
            a_read = A["a"]
        self.assertEqual(a, a_read)
        

    def test_mpi(self):

        a=Toto(0)

        if mpi.is_master_node():
            a=Toto(1)
            mpi.bcast(a)

        self.assertEqual(a, Toto(1))

class test_chain(unittest.TestCase):

    def test_chain(self):

        i = 111
        j = 222
        ij = chain(i,j)
        self.assertEqual(ij, 111222)

if __name__ == '__main__':
    unittest.main()


test/c++/anderson_c.cpp
#include <ctint_tutorial/solver.hpp>

#include <triqs/gfs.hpp>
#include <triqs/test_tools/gfs.hpp>

using namespace triqs::gfs; 
using namespace ctint_tutorial; 

// Anderson model test
TEST(CtInt, Anderson) {

  // Initialize mpi
  int rank = mpi::communicator().rank();

  // set parameters
  double beta  = 20.0;
  double U     = 1.0;
  double delta = 0.1;
  int n_cycles = 10000;
  int n_iw     = 200;

  // construct the ct-int solver
  auto ctqmc = solver{beta, n_iw};

  // parameters
  double mu      = 1.3 - U / 2.0; // mu=0 corresponds to half filling
  double epsilon = 0.2;

  // initialize g0(omega)
  triqs::clef::placeholder<0> om_;
  for (auto sig : {0, 1}) ctqmc.G0_iw()[sig](om_) << 1.0 / (om_ + mu - 1.0 / (om_ - epsilon));

  // launch the Monte Carlo
  ctqmc.solve(U, delta, n_cycles);

  // to compare with ct_seg
  // gf<imfreq> gw = ctqmc.G0_iw()[0];
  // auto gt = make_gf_from_inverse_fourier(gw);

  std::string filename = "anderson_c";
  gf<imfreq> g;
  if (rank == 0) {
    triqs::h5::file G_file(filename + ".ref.h5", 'r');
    h5_read(G_file, "G", g);
    EXPECT_GF_NEAR(g, ctqmc.G_iw()[0]);
  }
  if (rank == 0) {
    triqs::h5::file G_file(filename + ".out.h5", 'w');
    h5_write(G_file, "G", ctqmc.G_iw()[0]);
  }
}
MAKE_MAIN;


examples/dmft_bethe.py
from pytriqs.gf import *
from pytriqs.archive import *
from pytriqs.plot.mpl_interface import oplot
import pytriqs.utility.mpi as mpi

from ctint_tutorial import Solver

# Parameters
U = 2.5            # Hubbard interaction
mu = U/2.0         # Chemical potential
half_bandwidth=1.0 # Half bandwidth (energy unit)
beta = 40.0        # Inverse temperature
n_iw = 128         # Number of Matsubara frequencies
n_cycles = 10000   # Number of MC cycles
delta = 0.1        # delta parameter
n_iterations = 1   # Number of DMFT iterations

S = Solver(beta, n_iw) # Initialize the solver

S.G_iw << SemiCircular(half_bandwidth) # Initialize the Green's function

with HDFArchive("dmft_bethe.output.h5",'w') as A:
 A['n_iterations'] = n_iterations # Save a parameter

 for it in range(n_iterations): # DMFT loop
  for name, G0 in S.G0_iw:
   G0 << inverse(iOmega_n + mu - (half_bandwidth/2.0)**2 * S.G_iw[name] ) # Set G0
  # Change random number generator on final iteration
  random_name = 'mt19937' if it<n_iterations-1 else 'lagged_fibonacci19937'

  S.solve(U, delta, n_cycles, random_name=random_name) # Solve the impurity problem

  G_sym = (S.G_iw['up']+S.G_iw['down'])/2 # Impose paramagnetic solution
  S.G_iw << G_sym

  if mpi.is_master_node():
   A['G%i'%it] = G_sym # Save G from every iteration to file


c++/ctint_tutorial/solver.cpp
#include "solver.hpp"

#include <triqs/mc_tools.hpp>
#include <triqs/det_manip.hpp>
#include <mpi/mpi.hpp>

using namespace triqs::gfs; 
using namespace triqs::arrays; 
using namespace ctint_tutorial;

// --------------- The QMC configuration ----------------

// Argument type of g0bar
struct arg_t {
  double tau; // The imaginary time
  int s;      // The auxiliary spin
};

// The function that appears in the calculation of the determinant
struct g0bar_tau {
  gf<imtime> const &gt;
  double beta, delta;
  int s;

  dcomplex operator()(arg_t const &x, arg_t const &y) const {
    if ((x.tau == y.tau)) { // G_\sigma(0^-) - \alpha(\sigma s)
      return 1.0 + gt[0](0, 0) - (0.5 + (2 * (s == x.s ? 1 : 0) - 1) * delta);
    }
    auto x_y = x.tau - y.tau;
    bool b   = (x_y >= 0);
    if (!b) x_y += beta;
    dcomplex res = gt[closest_mesh_pt(x_y)](0, 0);
    return (b ? res : -res); // take into account antiperiodicity
  }
};

// The Monte Carlo configuration
struct configuration {
  // M-matrices for up and down
  std::vector<triqs::det_manip::det_manip<g0bar_tau>> Mmatrices;

  int perturbation_order() const { return Mmatrices[up].size(); }

  configuration(block_gf<imtime> &g0tilde_tau, double beta, double delta) {
    // Initialize the M-matrices. 100 is the initial matrix size
    for (auto spin : {up, down}) Mmatrices.emplace_back(g0bar_tau{g0tilde_tau[spin], beta, delta, spin}, 100);
  }
};

// ------------ QMC move : inserting a vertex ------------------

struct move_insert {
  configuration *config;
  triqs::mc_tools::random_generator &rng;
  double beta, U;

  dcomplex attempt() { // Insert an interaction vertex at time tau with aux spin s
    double tau     = rng(beta);
    int s          = rng(2);
    auto k         = config->perturbation_order();
    auto det_ratio = config->Mmatrices[up].try_insert(k, k, {tau, s}, {tau, s}) * config->Mmatrices[down].try_insert(k, k, {tau, s}, {tau, s});
    return -beta * U / (k + 1) * det_ratio; // The Metropolis ratio
  }

  dcomplex accept() {
    for (auto &d : config->Mmatrices) d.complete_operation(); // Finish insertion
    return 1.0;
  }

  void reject() {
    for (auto &d : config->Mmatrices) d.reject_last_try(); // Finish insertion
  }
};

// ------------ QMC move : deleting a vertex ------------------

struct move_remove {
  configuration *config;
  triqs::mc_tools::random_generator &rng;
  double beta, U;

  dcomplex attempt() {
    auto k = config->perturbation_order();
    if (k <= 0) return 0;    // Config is empty, trying to remove makes no sense
    int p          = rng(k); // Choose one of the operators for removal
    auto det_ratio = config->Mmatrices[up].try_remove(p, p) * config->Mmatrices[down].try_remove(p, p);
    return -k / (beta * U) * det_ratio; // The Metropolis ratio
  }

  dcomplex accept() {
    for (auto &d : config->Mmatrices) d.complete_operation();
    return 1.0;
  }

  void reject() {
    for (auto &d : config->Mmatrices) d.reject_last_try(); // Finish insertion
  }                                                        // Nothing to do
};

//  -------------- QMC measurement ----------------

struct measure_M {

  configuration const *config; // Pointer to the MC configuration
  block_gf<imfreq> &Mw;        // reference to M-matrix
  double beta;
  dcomplex Z = 0;
  long count = 0;

  measure_M(configuration const *config_, block_gf<imfreq> &Mw_, double beta_) : config(config_), Mw(Mw_), beta(beta_) { Mw() = 0; }

  void accumulate(dcomplex sign) {
    Z += sign;
    count++;

    for (auto spin : {up, down}) {

      // A lambda to measure the M-matrix in frequency
      auto lambda = [this, spin, sign](arg_t const &x, arg_t const &y, dcomplex M) {
        auto const &mesh = this->Mw[spin].mesh();
        auto phase_step  = -1_j * M_PI * (x.tau - y.tau) / beta;
        auto coeff       = std::exp((2 * mesh.first_index() + 1) * phase_step);
        auto fact        = std::exp(2 * phase_step);
        for (auto const &om : mesh) {
          this->Mw[spin][om](0, 0) += sign * M * coeff;
          coeff *= fact;
        }
      };

      foreach (config->Mmatrices[spin], lambda)
        ;
    }
  }

  void collect_results(mpi::communicator const &c) {
    Mw = mpi::all_reduce(Mw, c);
    Z  = mpi::all_reduce(Z, c);
    Mw = Mw / (-Z * beta);

    // Print the sign
    if (c.rank() == 0) std::cerr << "Average sign " << Z / c.size() / count << std::endl;
  }
};

// ------------ The main class of the solver ------------------------

solver::solver(double beta_, int n_iw, int n_tau)
   : beta{beta_},
     g0_iw{make_block_gf({"up", "down"}, gf<imfreq>{{beta, Fermion, n_iw}, {1, 1}})},
     g0tilde_iw{g0_iw},
     g_iw{g0_iw},
     M_iw{g0_iw},
     g0tilde_tau{make_block_gf({"up", "down"}, gf<imtime>{{beta, Fermion, n_tau}, {1, 1}})} {}

// The method that runs the qmc
void solver::solve(double U, double delta, int n_cycles, int length_cycle, int n_warmup_cycles, std::string random_name, int max_time) {

  mpi::communicator world;
  triqs::clef::placeholder<0> spin_;
  triqs::clef::placeholder<1> om_;

  for (auto spin : {up, down}) { // Apply shift to g0_iw and Fourier transform
    g0tilde_iw[spin](om_) << 1.0 / (1.0 / g0_iw[spin](om_) - U / 2);
    array<dcomplex, 3> mom{{{0}}, {{1}}}; // Fix the moments: 0 + 1/omega
    g0tilde_tau()[spin] = triqs::gfs::fourier(g0tilde_iw[spin], make_const_view(mom));
  }

  // Rank-specific variables
  int verbosity   = (world.rank() == 0 ? 3 : 0);
  int random_seed = 34788 + 928374 * world.rank();

  // Construct a Monte Carlo loop
  triqs::mc_tools::mc_generic<dcomplex> CTQMC(random_name, random_seed, verbosity);

  // Prepare the configuration
  auto config = configuration{g0tilde_tau, beta, delta};

  // Register moves and measurements
  CTQMC.add_move(move_insert{&config, CTQMC.get_rng(), beta, U}, "insertion");
  CTQMC.add_move(move_remove{&config, CTQMC.get_rng(), beta, U}, "removal");
  CTQMC.add_measure(measure_M{&config, M_iw, beta}, "M measurement");

  // Run and collect results
  CTQMC.warmup_and_accumulate(n_warmup_cycles, n_cycles, length_cycle, triqs::utility::clock_callback(max_time));
  CTQMC.collect_results(world);

  // Compute the Green function from Mw
  g_iw[spin_](om_) << g0tilde_iw[spin_](om_) + g0tilde_iw[spin_](om_) * M_iw[spin_](om_) * g0tilde_iw[spin_](om_);

  // Set the tail of g_iw to 1/w
  triqs::arrays::array<dcomplex, 3> mom{{{0}}, {{1}}}; // 0 + 1/omega
  for (auto &g : g_iw) replace_by_tail_in_fit_window(g(), make_const_view(mom));
}


c++/ctint_tutorial/solver.hpp
#include <triqs/gfs.hpp>

// ------------ The main class of the solver -----------------------

namespace ctint_tutorial {

  enum spin { up, down };

  using namespace triqs::gfs; 

  class solver {

    double beta;
    triqs::gfs::block_gf<imfreq> g0_iw, g0tilde_iw, g_iw, M_iw;
    triqs::gfs::block_gf<imtime> g0tilde_tau;

    public:
    /// Access non-interacting Matsubara Green function
    block_gf_view<imfreq> G0_iw() { return g0_iw; }

    /// Access non-interacting imaginary-time Green function
    block_gf_view<imtime> G0_tau() { return g0tilde_tau; }

    /// Access interacting Matsubara Green function
    block_gf_view<imfreq> G_iw() { return g_iw; }

    /// Construct a ctint solver
    solver(double beta_, int n_iw = 1024, int n_tau = 100001);

    /// Method that performs the QMC calculation
    void solve(double U, double delta, int n_cycles, int length_cycle = 50, int n_warmup_cycles = 5000, std::string random_name = "",
               int max_time = -1);
  };
} // namespace ctint_tutorial


python/ctint_tutorial/__init__.py
################################################################################
#
# TRIQS: a Toolbox for Research in Interacting Quantum Systems
#
# Copyright (C) 2016-2018, N. Wentzell
# Copyright (C) 2018-2019, The Simons Foundation
#   author: N. Wentzell
#
# TRIQS is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# TRIQS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# TRIQS. If not, see <http://www.gnu.org/licenses/>.
#
################################################################################

r"""
DOC

"""
from ctint_tutorial_module import Solver

__all__ = ['Solver']


c++/ctint_tutorial/ctint_tutorial.cpp
#include <cmath>
#include "./ctint_tutorial.hpp"

namespace ctint_tutorial {

  toto &toto::operator+=(toto const &b) {
    this->i += b.i;
    return *this;
  }

  toto toto::operator+(toto const &b) const {
    auto res = *this;
    res += b;
    return res;
  }

  bool toto::operator==(toto const &b) const { return (this->i == b.i); }

  void h5_write(triqs::h5::group grp, std::string subgroup_name, toto const &m) {
    grp = subgroup_name.empty() ? grp : grp.create_group(subgroup_name);
    h5_write(grp, "i", m.i);
    h5_write_attribute(grp, "TRIQS_HDF5_data_scheme", toto::hdf5_scheme());
  }

  void h5_read(triqs::h5::group grp, std::string subgroup_name, toto &m) {
    grp = subgroup_name.empty() ? grp : grp.open_group(subgroup_name);
    int i;
    h5_read(grp, "i", i);
    m = toto(i);
  }

  int chain(int i, int j) {
    int n_digits_j = j > 0 ? (int)log10(j) + 1 : 1;
    return i * int(pow(10, n_digits_j)) + j;
  }

} // namespace ctint_tutorial


c++/ctint_tutorial/ctint_tutorial.hpp
#pragma once
#include <triqs/gfs.hpp>

namespace ctint_tutorial {

  /**
   * A very useful and important class
   *
   * @note A Useful note
   * @include ctint_tutorial/ctint_tutorial.hpp
   */
  class toto {

    int i = 0;

    public:
    toto() = default;

    /**
     * Construct from integer
     *
     * @param i_ a scalar  :math:`G(\tau)`
     */
    explicit toto(int i_) : i(i_) {}

    ~toto() = default;

    // Copy/Move construction
    toto(toto const &) = default;
    toto(toto &&)      = default;

    /// Copy/Move assignment
    toto &operator=(toto const &) = default;
    toto &operator=(toto &&) = default;

    /// Simple accessor
    [[nodiscard]] int get_i() const { return i; }

    /** 
     * A simple function with :math:`G(\tau)`
     *
     * @param u Nothing useful
     */
    int f(int u) { return u; }

    /// Arithmetic operations
    toto operator+(toto const &b) const;
    toto &operator+=(toto const &b);

    /// Comparison
    bool operator==(toto const &b) const;

    /// HDF5
    static std::string hdf5_scheme() { return "Toto"; }

    friend void h5_write(triqs::h5::group grp, std::string subgroup_name, toto const &m);
    friend void h5_read(triqs::h5::group grp, std::string subgroup_name, toto &m);

    /// Serialization
    template <class Archive> void serialize(Archive &ar, const unsigned int) { ar &i; }
  };

  /**
   * Chain digits of two integers
   *
   * @head A set of functions that implement chaining
   *
   * @tail Do I really need to explain more ? 
   *
   * @param i The first integer
   * @param j The second integer
   * @return An integer containing the digits of both i and j
   *
   */
  int chain(int i, int j);

} // namespace ctint_tutorial


python/ctint_tutorial/ctint_tutorial_module_desc.py
# Generated automatically using the command :
# c++2py ../../c++/ctint_tutorial/solver.hpp -p --members_read_only -N ctint_tutorial -a ctint_tutorial -m ctint_tutorial_module -o ctint_tutorial_module --moduledoc="The ctint_tutorial python module" -C pytriqs --cxxflags="-std=c++17" --target_file_only
from cpp2py.wrap_generator import *

# The module
module = module_(full_name = "ctint_tutorial_module", doc = r"The ctint_tutorial python module", app_name = "ctint_tutorial")

# Imports
module.add_imports(*['pytriqs.gf'])

# Add here all includes
module.add_include("ctint_tutorial/solver.hpp")

# Add here anything to add in the C++ code at the start, e.g. namespace using
module.add_preamble("""
#include <cpp2py/converters/string.hpp>
#include <triqs/cpp2py_converters/gf.hpp>

using namespace ctint_tutorial;
""")

module.add_enum("spin", ['spin::up', 'spin::down'], "ctint_tutorial", doc = r"""""")

# The class solver
c = class_(
        py_type = "Solver",  # name of the python class
        c_type = "ctint_tutorial::solver",   # name of the C++ class
        doc = r"""""",   # doc of the C++ class
        hdf5 = False,
)

c.add_constructor("""(double beta_, int n_iw = 1024, int n_tau = 100001)""", doc = r"""Construct a ctint solver""")

c.add_method("""void solve (double U, double delta, int n_cycles, int length_cycle = 50, int n_warmup_cycles = 5000, std::string random_name = \"\", int max_time = -1)""",
             doc = r"""Method that performs the QMC calculation""")

c.add_property(name = "G0_iw",
               getter = cfunction("block_gf_view<triqs::gfs::imfreq> G0_iw ()"),
               doc = r"""Access non-interacting Matsubara Green function""")

c.add_property(name = "G0_tau",
               getter = cfunction("block_gf_view<triqs::gfs::imtime> G0_tau ()"),
               doc = r"""Access non-interacting imaginary-time Green function""")

c.add_property(name = "G_iw",
               getter = cfunction("block_gf_view<triqs::gfs::imfreq> G_iw ()"),
               doc = r"""Access interacting Matsubara Green function""")

module.add_class(c)



module.generate_code()